New pip build system

dbort opened the pull request:
Summary:
Compatible with `python -m build -wheel` instead of using the deprecated `python setup.py bdist_wheel`.

A new custom build command builds all CMake targets, and a new custom build_ext command installs built files into the wheel.

A new custom build_py command copies platform-independent files into the wheel instead of copying them into the source tree.

Puts all generated files under `pip-out` instead of mixing with files in the tree.

Uses Command methods like `mkpath` and `copy_file` to benefit from setuptools concepts like dry_run, and to get logging for free.

Adds `flatc` to the wheel and puts it on the user's PATH. Removes the call to `install_flatc.sh` in `install_requirements.sh`; but, leaves the script in place since someone might want to run it on their own.

Also:
- Enables runtime logging for pybindings even in release mode
- Fixes another `-fPIC` requirement when building the `flatcc` host tools
- Picks a number of build jobs based on the cores in the system, instead of defaulting to 9
- Removes now-unnecessary steps from the Getting Started docs
- Removes a use of `${arg^^}`, which doesn't work with older versions of bash

Test Plan:
Iteration command to test wheel building without completely rebuilding the system:
```
rm -rf \
    /tmp/wheel-out \
    pip-out/lib* \
    pip-out/bdist.* \
    pip-out/temp.*/cmake-out/CMake* \
    executorch.egg-info \
    third-party/flatcc/{bin,lib} \
    ; \
  python setup.py bdist_wheel --dist-dir /tmp/wheel-out
```
To fully rebuild the system, also `rm -rf pip-out`.

The wheel file is written to (e.g.)
```
/tmp/wheel-out/executorch-0.1.0-cp310-cp310-linux_x86_64.whl
```

Examine it with `unzip -l`, or install it with `pip install /tmp/wheel-out/executorch*.whl`.

Also tested creating the wheel via `install_requirements.sh` and creating a simple model file.
```
./install_requirements.sh
python3 -m examples.portable.scripts.export --model_name="add"
<creates add.pte file>
```

Also tested that it can build pybindings with xnnpack when invoked via `install_requirements.sh`:
```
./install_requirements.sh --pybind xnnpack
python -m unittest backends/xnnpack/test/models/mobilenet_v2.py
```
Builds successfully, but trying to run the test currently fails during xnnpack init() due to T184024365. But it fails in the same way that the main branch does, so this new diff does not regress.

Reviewers:

Subscribers:

Tasks:

Tags:

dbort commented:
I still need to clean this up, but now that it's working I'd like to get some eyes on it. There's a lot going on here that's hard to separate, since it's a rewrite of setup.py.

larryliu0820 commented on .gitignore:
```diff
@@ -5,12 +5,8 @@ cmake-android-out/
+pip-out/
 __pycache__/
-build/lib/
-exir/_serialize/scalar_type.fbs
-exir/_serialize/program.fbs
-sdk/bundled_program/serialize/bundled_program_schema.fbs
-sdk/bundled_program/serialize/scalar_type.fbs
```
Oh nice finding we probably don't want to ignore all these

dbort replied:
We needed to before this, because setup.py copied them directly into the source tree.

larryliu0820 commented on CMakeLists.txt:
```diff
@@ -301,10 +301,10 @@ endif()
 install(
   TARGETS executorch
-  DESTINATION lib
+  LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
```
Does CMake provide a default value for `CMAKE_INSTALL_LIBDIR` or we need to always pass in a value

dbort replied:
Yeah it defaults to `lib` but can be `lib64` in some situations, or a big ProgramFiles path on windows. So this is more portable then just saying `lib`.

larryliu0820 commented on install_requirements.sh:
```diff
@@ -69,6 +69,8 @@ pip install --force-reinstall --pre transformers==${TRANSFORMERS_VERSION}
 TORCHSR_VERSION=1.0.4
 pip install --pre torchsr==${TORCHSR_VERSION}
 
+echo STOP # @nocommit
+exit 1
```
remove this?

larryliu0820 commented on pyproject.toml:
```diff
@@ -1,5 +1,13 @@
+requires = [
+  "setuptools",
+  "wheel",
+# @nocommit: insted we're using --no-isolation
```
Yeah is there a way to force `--no-build-isolation` in pip install?

larryliu0820 commented on setup.py:
```diff
@@ -46,194 +46,332 @@
+        #
+        # See build.initialize_options() in
+        # setuptools/_distutils/command/build.py for the default.
+        self.build_base = "pip-out"
```
so one thing I noticed in the previous pip install experience, is that the build base directory name contains useful information such as python version and os arch. Do we still have that in the new directory name?

dbort replied:
Yeah, this just moves the root from `build/...` to `pip-out/...` so it doesn't sit on top of our existing `build/` directory. All of the platform-based directories live under this.

larryliu0820 reviewed (commented):
Thanks for putting this up! It would be great if you can paste the temp directory tree structure and also things in the wheel so that we know what's inside.

dbort commented:
> It would be great if you can paste the temp directory tree structure and also things in the wheel so that we know what's inside.

Here's the listing from the latest version: https://gist.github.com/dbort/4a79eac7e1d1ebe7306f437bb2583482

The `data` directory is the interesting part that I've been getting into shape. Now you can see that it has `flatc`, and an __init__.py wrapper that knows how to execute it. See https://github.com/pytorch/executorch/pull/2499/commits/9b5e45409d4b0d64bb2e182b5df20da9d054449a

```
     1811  03-23-2024 02:33   executorch/data/bin/__init__.py
  2250288  03-23-2024 01:35   executorch/data/bin/flatc
   179456  03-23-2024 01:38   executorch/data/lib/liboptimized_native_cpu_ops_lib.a
    16414  03-23-2024 01:35   executorch/data/lib64/libbundled_program.a
    41846  03-23-2024 01:35   executorch/data/lib64/libetdump.a
   174748  03-23-2024 02:42   executorch/data/lib64/libexecutorch.a
    12056  03-23-2024 01:35   executorch/data/lib64/libextension_data_loader.a
   203986  03-23-2024 02:42   executorch/data/lib64/libflatccrt.a
 88238252  03-23-2024 01:38   executorch/data/lib64/libportable_kernels.a
   179392  03-23-2024 01:38   executorch/data/lib64/libportable_ops_lib.a
     2715  03-23-2024 01:34   executorch/data/lib64/cmake/executorch/executorch-config.cmake
```

Oh whoops looks like I need to update liboptimized_native_cpu_ops_lib.a to use the proper LIBDIR. (lib64 is the LIBDIR value picked by the GNU Coding Conventions when building on my x86_64 machine)

larryliu0820 reviewed (approved):
Not sure if ci is reliable at this stage..

dbort committed:
New pip build system

Summary:
Compatible with `pythom -m build` instead of using the deprecated
`python setup.py bdist_wheel`.

A new custom build command builds all necessary CMake targets, and a new
custom build_ext command installs built files into the wheel.

A new custom build_py command copies platform-independent files into the
wheel instead of copying them into the source tree.

Puts all generated files under pip-out instead of mixing with files in
the tree.

Uses Command methods like `mkpath` and `copy_file` to benefit from
setuptools concepts like dry_run, and to get logging for free.

Adds `flatc` to the wheel and installs a wrapper to make it available
on the path.

Test Plan:
Iteration command to test wheel building without completely rebuilding
the system:
```
rm -rf \
    /tmp/wheel-out \
    pip-out/lib* \
    pip-out/bdist.* \
    pip-out/temp.*/cmake-out/CMake* \
    executorch.egg-info \
    third-party/flatcc/{bin,lib} \
    ; \
  python setup.py bdist_wheel --dist-dir /tmp/wheel-out
```
To fully rebuild the system, also `rm -rf pip-out`.

The wheel file is written to (e.g.)
```
/tmp/wheel-out/executorch-0.1.0-cp310-cp310-linux_x86_64.whl
```

Examine it with `unzip -l`, or install it with `pip install
/tmp/wheel-out/executorch*.whl`.

GregoryComer commented on setup.py:
```diff
@@ -47,194 +47,390 @@
+
+        # Default build parallelism based on number of cores, but allow
+        # overriding through the environment.
+        default_parallel = str(os.cpu_count() - 1)
```
Nice change. This should speed up builds significantly on high core count systems.

GregoryComer commented:
This looks great. Thanks for taking all of this on.
